File: ganapathy_pavers/ganapathy_pavers/ganapathy_pavers/report/job_worker___report/job_worker___report.py
# Copyright (c) 2022, Thirvusoft and contributors
# For license information, please see license.txt

import frappe
from frappe import _
def execute(filters=None):
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    employee = filters.get("employee")
    site_name = filters.get("site_name")
    conditions = ""
    adv_conditions = ""
    if from_date or to_date or employee or site_name:
        conditions = " where 1 = 1"
        adv_conditions = "where 1 = 1"
        if from_date and to_date:
            conditions += "  and jwd.start_date between '{0}' and '{1}' ".format(from_date, to_date)
            adv_conditions += " and empadv.posting_date between '{0}' and '{1}' ".format(from_date, to_date)
        if employee:
            conditions += " and jwd.name1 ='{0}' ".format(employee)
        if site_name:
            conditions += " and site.name = '{0}' ".format(site_name)
    report_data = frappe.db.sql(""" select *,(amount + salary_balance - advance_amount) from (select (select employee_name from `tabEmployee` where name = jwd.name1 ) as jobworker,site.name,site.status,jwd.sqft_allocated,
                                        emp.salary_balance as salary_balance,jwd.amount as amount,
                                        (select sum(empadv.advance_amount - empadv.return_amount) from `tabEmployee Advance` as empadv {1} and empadv.employee = jwd.name1 and docstatus = 1) as advance_amount
                                        from `tabProject` as site
                                        left outer join `tabTS Job Worker Details` as jwd
                                            on site.name = jwd.parent
                                        left outer join `tabEmployee` as emp
                                            on emp.employee = jwd.name1
                                        {0}
                                    group by jwd.name1,jwd.sqft_allocated)as total_cal
                                """.format(conditions,adv_conditions))
    data = [list(i) for i in report_data]
    final_data = []
    c = 0
    
    if(len(data)):
        start = 0
        for i in range(len(data)-1):
            if (data[i][0] != data[i+1][0]):
                adv=data[i][6]
                data[i][6]=[]
                final_data.append(data[i])
                total = [" " for i in range(8)]
                total[2] = "<b style=color:orange;>""Total""</b>"
                total[3] = sum((data[i][3] or 0) for i in range(start,i+1))
                total[4] = sum((data[i][4] or 0) for i in range(start,i+1))
                total[5] = sum((data[i][5] or 0) for i in range(start,i+1))
                total[6] = adv
                total[7] = sum((data[i][7] or 0) for i in range(start,i+1))
                final_data.append(total)
                start = i+1	
                c=0
            else:
                if(c==0):data[i][6]=()
                else:data[i][6]=()
                c+=1
                final_data.append(data[i])
        adv=data[-1][6]
        data[-1][6]=()      
        final_data.append(data[-1])
        total = [" " for i in range(8)]
        total[2] = "<b style=color:orange;>""Total""</b>"
        total[3] = sum((data[i][3] or 0) for i in range(start,len(data)))
        total[4] = sum((data[i][4] or 0) for i in range(start,len(data)))
        total[5] = sum((data[i][5] or 0) for i in range(start,len(data)))
        total[6] = adv
        total[7] = sum((data[i][7] or 0) for i in range(start,len(data)))
        final_data.append(total)
    columns = get_columns()
    return columns, final_data

def get_columns():
	columns = [
		_("Job Worker") + ":Data/Employee:150",
		_("Site Name") + ":Link/Project:150",
		_("Status") + ":Data/Project:150",
		_("Completed Sqft") + ":Data:150",
		_("Salary Balance") + ":Currency:150",
		_("Amount") + ":Currency:150",
		_("Advance Amount") + ":Currency:150",
		_("Total Amount") + ":Currency:150",
		]
	
	return columns

File: ganapathy_pavers/ganapathy_pavers/report/job_worker___report/job_worker___report.js
// Copyright (c) 2022, Thirvusoft and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Job Worker - Report"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_days(frappe.datetime.get_today(), -7),
			"width": "80",
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_days(frappe.datetime.get_today(), -1),
			"width": "80",
			"reqd": 1
		},
		{
			"fieldname": "site_name",
			"label": __("Site Name"),
			"fieldtype": "Link",
			"options": "Project",
			"width": "100"
		},
		{
			"fieldname": "employee",
			"label": __("Job Worker"),
			"fieldtype": "Link",
			"options": "Employee",
			"width": "100",
			get_query: function() {
				return {
					filters: {
						designation: "Job Worker"
					}
				}
			}
		},
		{
			"fieldname": "type",
			"label": __("Type"),
			"fieldtype": "Select",
			"options": "\nPavers\nCompound Wall"
		},
		{
			"fieldname": "payment_date",
			"label": __("Payment Date"),
			"fieldtype": "Date",
			on_change: function () { }
		},
		{
			"fieldname": "group_site_work",
			"label": __("Group Site Work"),
			"fieldtype": "Check",
			"default": 1
		},
		{
			"fieldname": "show_only_other_work",
			"label": __("Show Only Other Work"),
			"fieldtype": "Check",
			on_change: function () {
				if (frappe.query_report.get_filter('show_only_other_work').get_value()) {
					frappe.query_report.get_filter('hide_other_work').set_value(!frappe.query_report.get_filter('show_only_other_work').get_value())
				}
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "hide_other_work",
			"label": __("Don't Show Other Work"),
			"fieldtype": "Check",
			on_change: function () {
				if (frappe.query_report.get_filter('hide_other_work').get_value()) {
					frappe.query_report.get_filter('show_only_other_work').set_value(!frappe.query_report.get_filter('hide_other_work').get_value())
				}
				frappe.query_report.refresh();
			}
		},
	],
	"formatter": function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (data && (data.status == "Total" || data.job_worker == "Total" || data.bold)) {
			value = `<b style="font-size: 117%;${(column.fieldname=="job_worker" && data.job_worker!="Total")?'font-weight:bold;font-size: 20px;text-align: center; color:green':''}">${value}</b>`;
		}
		if (data.status == 'DETAILS') {
			value = `<b style="font-size: 117%;${(column.fieldname=="job_worker")?'font-weight:bold;font-size: 20px;text-align: center; color:green':'display:none'}">${value}</b>`;
		}
		return value;
	},
};


frappe.query_report.print_report = function print_report(print_settings) {
	const custom_format = frappe.query_report.report_settings.html_format || null;
	const filters_html = frappe.query_report.get_filters_html_for_print();
	const landscape = print_settings.orientation == 'Landscape';

	frappe.query_report.make_access_log('Print', 'PDF');
	frappe.render_grid({
		template: /*print_settings.columns */!custom_format ? 'print_grid' : custom_format,
		title: __(frappe.query_report.report_name),
		subtitle: filters_html,
		print_settings: print_settings,
		landscape: landscape,
		filters: frappe.query_report.get_filter_values(),
		data: frappe.query_report.get_data_for_print(),
		columns: frappe.query_report.get_columns_for_print(print_settings, custom_format),
		original_data: frappe.query_report.data,
		report: frappe.query_report
	});
}

frappe.query_report.pdf_report = function pdf_report(print_settings) {
	const base_url = frappe.urllib.get_base_url();
	const print_css = frappe.boot.print_css;
	const landscape = print_settings.orientation == 'Landscape';

	const custom_format = frappe.query_report.report_settings.html_format || null;
	const columns = frappe.query_report.get_columns_for_print(print_settings, custom_format);
	const data = frappe.query_report.get_data_for_print();
	const applied_filters = frappe.query_report.get_filter_values();

	const filters_html = frappe.query_report.get_filters_html_for_print();
	const template =
		/*print_settings.columns ||*/ !custom_format ? 'print_grid' : custom_format;
	const content = frappe.render_template(template, {
		title: __(frappe.query_report.report_name),
		subtitle: filters_html,
		filters: applied_filters,
		data: data,
		original_data: frappe.query_report.data,
		columns: columns,
		report: frappe.query_report
	});

	// Render Report in HTML
	const html = frappe.render_template('print_template', {
		title: __(frappe.query_report.report_name),
		content: content,
		base_url: base_url,
		print_css: print_css,
		print_settings: print_settings,
		landscape: landscape,
		columns: columns,
		lang: frappe.boot.lang,
		layout_direction: frappe.utils.is_rtl() ? "rtl" : "ltr"
	});

	frappe.render_pdf(html, print_settings);
}